src/rules/deprecationRule.ts

```diff
@@ -0,0 +1,80 @@
+/**
+ * @license
+ * Copyright 2016 Palantir Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as ts from "typescript";
+import * as Lint from "../index";
+
+export class Rule extends Lint.Rules.TypedRule {
+
+    /* tslint:disable:object-literal-sort-keys */
+    public static metadata: Lint.IRuleMetadata = {
+        ruleName: "deprecation",
+        description: "Warns when deprecated APIs are used.",
+        descriptionDetails: Lint.Utils.dedent`Any usage of an identifier
+            with the @deprecated JSDoc annotation will trigger a warning.
+            See http://usejsdoc.org/tags-deprecated.html`,
+        rationale: Lint.Utils.dedent`
+            Deprecated APIs should be avoided, and usage updated.`,
+        optionsDescription: "",
+        options: null,
+        optionExamples: [],
+        type: "maintainability",
+        typescriptOnly: false,
+    };
+    /* tslint:enable:object-literal-sort-keys */
+    public applyWithProgram(sourceFile: ts.SourceFile, program: ts.Program): Lint.RuleFailure[] {
+        return this.applyWithWalker(new Walker(sourceFile, this.getOptions(), program));
+    }
+}
+
+class Walker extends Lint.ProgramAwareRuleWalker {
+  // Implementation inspired by angular/tsickle:
+  // https://github.com/angular/tsickle/blob/cad7c180a2155db6f6fb8d22c44151d7e8a9149f/src/decorator-annotator.ts#L42
+  protected visitIdentifier(node: ts.Identifier) {
+    let decSym = this.getTypeChecker().getSymbolAtLocation(node);
```

Comment:
the metadata is only used to generate documentation, it doesn't influence the behavior of the rule

The typescript API you are using may return `undefined` even though its signature does not declare that. You already mentioned the problem with strictNullChecks in TS APIs below...